mixnet/src/utils.rs

```diff
@@ -1,20 +1,125 @@
 use std::time::Duration;
 
-use rand::thread_rng;
+use rand::{Rng, SeedableRng};
 use rand_distr::{Distribution, Exp};
 
-/// Get a random interval from an exponential distribution.
-pub fn poisson_interval_sec(rate_per_min: f64) -> Duration {
-    // create an exponential distribution
-    let exp = Exp::new(rate_per_min).unwrap();
-    // generate a random value from the distribution
-    let interval_min = exp.sample(&mut thread_rng());
-    // convert minutes to seconds
-    Duration::from_secs_f64(interval_min * 60.0)
+/// A struct which helps give us a random interval from an exponential distribution.
+#[derive(Copy, Clone, Debug)]
+pub struct PoissonExpInterval<R>(R);
+
+impl<R: SeedableRng> PoissonExpInterval<R> {
+    /// Returns a new instance
+    pub fn new(seed: R::Seed) -> Self {
+        Self(R::from_seed(seed))
+    }
 }
 
-/// Get the mean interval in seconds for a Poisson.
-pub fn poisson_mean_interval_sec(rate_per_min: f64) -> Duration {
-    // the mean interval in seconds for a Poisson process
-    Duration::from_secs_f64(1.0 / rate_per_min * 60.0)
+impl<R: Rng> PoissonExpInterval<R> {
+    /// Get a random interval from an exponential distribution in seconds.
+    ///
+    /// The Poisson (exponential) distribution is used because of its memoryless properties,
+    /// which is known to offer optimal anonymity properties (Nym whitepaper 4.4).
+    pub fn interval(&mut self, rate_per_min: f64) -> Duration {
```

Comment:
Why not generic function instead of a struct?

Reply:
To avoid always providing a `seed` to the function, I assume that we will invoke this function multiple times.

Reply:
We may have to change the randomness source as well. @youngjoonlee can you confirm?

Reply:
In our Mixnet specification, we don't strictly specify how randomness source has to be set for Poisson. 
Poisson distribution is used for both the interval of packet emissions (by msg senders) and the packet delays (by mix nodes). As long as all of them follow a Poisson distribution with the same "rate parameter" (aka. lambda) for a long time, the effect of mixing can be maintained as calculated in the [spec](https://www.notion.so/Mixnet-Specification-807b624444a54a4b88afa1cc80e100c2?pvs=4#535f72f7b3eb4577adfde7f4124fde4b). From my understanding, the randomness source is not very important as long as the same rate parameter is used across the network. Implementers (including us) can decide on the strategy whether they'll change the source periodically. It's okay even if the same randomness source is shared across the network.

Similarly, Nym implementation uses `OsRng`: [reference1](https://github.com/nymtech/nym/blob/develop/common/client-core/src/client/cover_traffic_stream.rs#L109-L109) and [reference2](https://github.com/nymtech/nym/blob/develop/common/client-core/src/client/real_messages_control/mod.rs#L147-L147). (they're still [discussing](https://github.com/nymtech/nym/blob/3748ab77a132143d5fd1cd75dd06334d33294815/common/nymsphinx/src/utils/mod.rs#L11) which Rng is better, though).

So, in our implementation, I think we can go with the struct because we're not gonna change the Rng until we find any better strategy. 
(Nym implementation has a generic function instead :) But I think both is okay.)

Reply:
@al8n @danielSanchezQ This PR has stayed for a long time. Now, I'm taking over this PR. We should try to make PRs not stay opened for too long.

I just added a final commit 80e68d23, which just defines a generic function instead of struct. While implementing mixnet logic, I've found we actually don't need `seed`. Let's go with the simplest structure, as Nym does: https://github.com/nymtech/nym/blob/release/v1.1.22/common/nymsphinx/src/utils/mod.rs. I think we don't need to make things complicated before knowing what we really need. 

@danielSanchezQ Please turn off the `change required` sign in this PR, so I can merge this PR.

Reply:
Also, renamed `utils.rs` to `poisson.rs`